bfdriver/efi/efi_vmexits.cpp

```diff
@@ -0,0 +1,382 @@
+    ::intel_x64::vmcs::guest_cs_access_rights::set(0x9B);
+
+    ::intel_x64::vmcs::guest_ds_selector::set(0);
+    ::intel_x64::vmcs::guest_ds_base::set(0);
+    ::intel_x64::vmcs::guest_ds_limit::set(0xFFFF);
+    ::intel_x64::vmcs::guest_ds_access_rights::set(0x93);
+
+    ::intel_x64::vmcs::guest_es_selector::set(0);
+    ::intel_x64::vmcs::guest_es_base::set(0);
+    ::intel_x64::vmcs::guest_es_limit::set(0xFFFF);
+    ::intel_x64::vmcs::guest_es_access_rights::set(0x93);
+
+    ::intel_x64::vmcs::guest_fs_selector::set(0);
+    ::intel_x64::vmcs::guest_fs_base::set(0);
+    ::intel_x64::vmcs::guest_fs_limit::set(0xFFFF);
+    ::intel_x64::vmcs::guest_fs_access_rights::set(0x93);
+
+    ::intel_x64::vmcs::guest_gs_selector::set(0);
+    ::intel_x64::vmcs::guest_gs_base::set(0);
+    ::intel_x64::vmcs::guest_gs_limit::set(0xFFFF);
+    ::intel_x64::vmcs::guest_gs_access_rights::set(0x93);
+
+    ::intel_x64::vmcs::guest_ss_selector::set(0);
+    ::intel_x64::vmcs::guest_ss_base::set(0);
+    ::intel_x64::vmcs::guest_ss_limit::set(0xFFFF);
+    ::intel_x64::vmcs::guest_ss_access_rights::set(0x93);
+
+    ::intel_x64::vmcs::guest_tr_selector::set(0);
+    ::intel_x64::vmcs::guest_tr_base::set(0);
+    ::intel_x64::vmcs::guest_tr_limit::set(0xFFFF);
+    ::intel_x64::vmcs::guest_tr_access_rights::set(0x8B);
+
+    ::intel_x64::vmcs::guest_ldtr_selector::set(0);
+    ::intel_x64::vmcs::guest_ldtr_base::set(0);
+    ::intel_x64::vmcs::guest_ldtr_limit::set(0xFFFF);
+    ::intel_x64::vmcs::guest_ldtr_access_rights::set(0x82);
+
+    ::intel_x64::vmcs::guest_gdtr_base::set(0);
+    ::intel_x64::vmcs::guest_gdtr_limit::set(0xFFFF);
+
+    ::intel_x64::vmcs::guest_idtr_base::set(0);
+    ::intel_x64::vmcs::guest_idtr_limit::set(0xFFFF);
+
+    vmcs->save_state()->rax = 0;
+    vmcs->save_state()->rbx = 0;
+    vmcs->save_state()->rcx = 0;
+    vmcs->save_state()->rdx = 0xF00;
+    vmcs->save_state()->rdi = 0;
+    vmcs->save_state()->rsi = 0;
+    vmcs->save_state()->rbp = 0;
+    vmcs->save_state()->rsp = 0;
+    vmcs->save_state()->rip = 0;
+
+    ::intel_x64::vmcs::guest_rflags::set(0x2);
+    ::intel_x64::vmcs::guest_ia32_efer::set(0);
+
+    ::intel_x64::vmcs::guest_activity_state::set(::intel_x64::vmcs::guest_activity_state::active);
+
+    return true;
+}
+
+void enable_microcode_update_passthrough()
+{
+
+    uint64_t target_msr = 0x79;
+    void *msr_bitmaps = bfvmm::memory_manager::instance()->alloc(0x1000);
+    memset(msr_bitmaps, 0xFF, 0x1000);
+
+    uint64_t msr_offset = target_msr / 8;
+    uint64_t write_offset = 2048;
+    uint8_t *addr = (uint8_t *)msr_bitmaps + write_offset + msr_offset;
+    *addr = *addr & ~(1 << (target_msr % 8));
+
+    ::intel_x64::vmcs::primary_processor_based_vm_execution_controls::use_msr_bitmap::enable();
+    ::intel_x64::vmcs::address_of_msr_bitmap::set((uint64_t)msr_bitmaps);
+
+}
+
+// -----------------------------------------------------------------------------
```

Comment:
Every thing from here down should be placed in bfvmm/include/vcpu/arch/intel_x64/\<efi-vcpu-filename\>

Reply:
to double check: bfvmm root folder or bfdriver root folder?